server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.routing.allocation.command.AllocationCommands;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+/**
+ * A service that resets allocation failures counter and triggers reroute. For
+ * example, reset counter when a new node joins a cluster. Resetting counter on
+ * new node join covers a variety of use cases, such as rolling update, version
+ * (and other settings), node restarts.
+ */
+public class AllocationFailuresResetService implements ClusterStateListener {
```

Comment:
Do we subscribe this listener to the cluster service somewhere? Sorry if I'm missing the line that does that. It'd be good to have an integ test to demonstrate that it is all wired in correctly as well as the unit tests that show that the service works in isolation.

Reply:
Will add integ test, it will catch missing subscription